autorest/azure/auth/auth.go

```diff
@@ -60,126 +75,181 @@ func NewAuthorizerFromEnvironment() (autorest.Authorizer, error) {
+	config.AADEndpoint = settings.Environment.ActiveDirectoryEndpoint
+	config.Resource = settings.Values[Resource]
+	return config, nil
+}
+
+// GetMSI creates a MSI config object from the available client ID.
+func (settings EnvironmentSettings) GetMSI() MSIConfig {
+	config := NewMSIConfig()
+	config.Resource = settings.Values[Resource]
+	config.ClientID = settings.Values[ClientID]
+	return config
+}
+
+// GetDeviceFlow creates a device-flow config object from the available client and tenant IDs.
+func (settings EnvironmentSettings) GetDeviceFlow() DeviceFlowConfig {
+	clientID, tenantID := settings.getClientAndTenant()
+	config := NewDeviceFlowConfig(clientID, tenantID)
+	config.AADEndpoint = settings.Environment.ActiveDirectoryEndpoint
+	config.Resource = settings.Values[Resource]
+	return config
+}
+
+// GetAuthorizer creates an Authorizer configured from environment variables in the order:
+// 1. Client credentials
+// 2. Client certificate
+// 3. Username password
+// 4. MSI
+func (settings EnvironmentSettings) GetAuthorizer() (autorest.Authorizer, error) {
 	//1.Client Credentials
-	if settings.clientSecret != "" {
-		config := NewClientCredentialsConfig(settings.clientID, settings.clientSecret, settings.tenantID)
-		config.AADEndpoint = settings.environment.ActiveDirectoryEndpoint
-		config.Resource = settings.resource
-		return config.Authorizer()
+	if c, e := settings.GetClientCredentials(); e == nil {
+		return c.Authorizer()
 	}
 
 	//2. Client Certificate
-	if settings.certificatePath != "" {
-		config := NewClientCertificateConfig(settings.certificatePath, settings.certificatePassword, settings.clientID, settings.tenantID)
-		config.AADEndpoint = settings.environment.ActiveDirectoryEndpoint
-		config.Resource = settings.resource
-		return config.Authorizer()
+	if c, e := settings.GetClientCertificate(); e == nil {
+		return c.Authorizer()
 	}
 
 	//3. Username Password
-	if settings.username != "" && settings.password != "" {
-		config := NewUsernamePasswordConfig(settings.username, settings.password, settings.clientID, settings.tenantID)
-		config.AADEndpoint = settings.environment.ActiveDirectoryEndpoint
-		config.Resource = settings.resource
-		return config.Authorizer()
+	if c, e := settings.GetUsernamePassword(); e == nil {
+		return c.Authorizer()
 	}
 
 	// 4. MSI
-	config := NewMSIConfig()
-	config.Resource = settings.resource
-	config.ClientID = settings.clientID
-	return config.Authorizer()
+	return settings.GetMSI().Authorizer()
 }
 
 // NewAuthorizerFromFile creates an Authorizer configured from a configuration file.
 func NewAuthorizerFromFile(baseURI string) (autorest.Authorizer, error) {
-	file, err := getAuthFile()
-	if err != nil {
-		return nil, err
-	}
-
-	resource, err := getResourceForToken(*file, baseURI)
+	settings, err := GetSettingsFromFile()
 	if err != nil {
 		return nil, err
 	}
-	return NewAuthorizerFromFileWithResource(resource)
+	return settings.ClientCredentialsAuthorizer(baseURI)
```

Comment:
I did add this, see `FileSettings.ClientCertificateAuthorizer()`

Reply:
I could update `NewAuthorizerFromFile()` to try `ClientCertificateAuthorizer()` if there are no client credentials (similar to how environment-based auth tries different things).  What do you think?

Reply:
yea i think its fine, I mean i think how we read it is dynamic so we can just check what is defined in them, since its invalid to have both I think. So only related question I have is should we return an error if both are missing or if both are defined ?

Reply:
IMO we mimic `NewAuthorizerFromEnvironment()`, so we do client credentials first and if that exists use it, else try client certificate, and if that fails (i.e. both entries are missing) then we fail.

Reply:
yea we can do that, I see your point. I would say its a slightly different scenario because in some sense I think the file format is invalid if both are present, but i'm fine with it either way. So your choice :)

Reply:
Pushed the change.